### ISSUE TITLE
1.10 kubelet doesn't start any daemonset containers with 1.12 control plane

### ISSUE THREAD

**Issue by alvaroaleman** [kubernetes/kubernetes#69346]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind feature

**What happened**:

I upgraded a kubernetes cluster with an existing 1.10 node to 1.12. From there on the kubelet didn't start any containers. Pods got scheduled and then went into the `MatchNodeSelector` status

**What you expected to happen**:#

The kubelet to start containers

**How to reproduce it (as minimally and precisely as possible)**:

Create a kubeadm 1.12 cluster, join a node with a 1.10 kubelet

**Anything else we need to know?**:

Bisecting revealed the patch that fixes this in the kubelet is b1a7f85c9e90ef84c51cd173a16e42ce2560fe35 which got introduced in #62453

cc @k82cn as you wrote the fixing patch

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/sig scheduling

**Comment by alvaroaleman** [comment]
Adding `--feature-gates=ScheduleDaemonSetPods=false` on both scheduler and controller-manager circumvents the issue.

Still, the new default breaks the Kubernetes version skew policy, maybe set it back to `false` for 1.12?

**Comment by k82cn** [comment]
any log to share for both scheduler and kubelet ?

**Comment by alvaroaleman** [comment]
@k82cn Yes ofc, here are the `v=9` logs of them:

* Scheduler: https://gist.github.com/alvaroaleman/a365f243436abd933cdff6d80d40effe
* Kubelet: https://gist.github.com/alvaroaleman/0335c6ecfa5eab02f35b7ba02085b41c

**Comment by alvaroaleman** [comment]
@ravisantoshgudimetla Here are the yamls of the node/pod:


<details>
  <summary>Node</summary>

```
$ k get node centos---1-10-7-hz-alvaro-5cb4687446-hhgd8 -o yaml
apiVersion: v1
kind: Node
metadata:
  annotations:
    node.alpha.kubernetes.io/ttl: "0"
    volumes.kubernetes.io/controller-managed-attach-detach: "true"
  creationTimestamp: 2018-10-05T19:33:54Z
  labels:
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/os: linux
    kubernetes.io/hostname: centos---1-10-7-hz-alvaro-5cb4687446-hhgd8
    machine-controller/owned-by: 432f3ce2-c8d5-11e8-bffe-9600001198f6
  name: centos---1-10-7-hz-alvaro-5cb4687446-hhgd8
  resourceVersion: "2338"
  selfLink: /api/v1/nodes/centos---1-10-7-hz-alvaro-5cb4687446-hhgd8
  uid: 986b3c4f-c8d5-11e8-bffe-9600001198f6
spec:
  externalID: centos---1-10-7-hz-alvaro-5cb4687446-hhgd8
  podCIDR: 10.244.5.0/24
  taints:
  - effect: NoSchedule
    key: node.kubernetes.io/not-ready
    timeAdded: 2018-10-05T19:33:54Z
status:
  addresses:
  - address: 159.69.177.178
    type: InternalIP
  - address: centos---1-10-7-hz-alvaro-5cb4687446-hhgd8
    type: Hostname
  allocatable:
    cpu: "1"
    ephemeral-storage: "18094252003"
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    memory: 1731172Ki
    pods: "110"
  capacity:
    cpu: "1"
    ephemeral-storage: 19633520Ki
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    memory: 1833572Ki
    pods: "110"
  conditions:
  - lastHeartbeatTime: 2018-10-05T19:35:14Z
    lastTransitionTime: 2018-10-05T19:33:54Z
    message: kubelet has sufficient disk space available
    reason: KubeletHasSufficientDisk
    status: "False"
    type: OutOfDisk
  - lastHeartbeatTime: 2018-10-05T19:35:14Z
    lastTransitionTime: 2018-10-05T19:33:54Z
    message: kubelet has sufficient memory available
    reason: KubeletHasSufficientMemory
    status: "False"
    type: MemoryPressure
  - lastHeartbeatTime: 2018-10-05T19:35:14Z
    lastTransitionTime: 2018-10-05T19:33:54Z
    message: kubelet has no disk pressure
    reason: KubeletHasNoDiskPressure
    status: "False"
    type: DiskPressure
  - lastHeartbeatTime: 2018-10-05T19:35:14Z
    lastTransitionTime: 2018-10-05T19:33:54Z
    message: kubelet has sufficient PID available
    reason: KubeletHasSufficientPID
    status: "False"
    type: PIDPressure
  - lastHeartbeatTime: 2018-10-05T19:35:14Z
    lastTransitionTime: 2018-10-05T19:33:54Z
    message: 'runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady
      message:docker: network plugin is not ready: cni config uninitialized'
    reason: KubeletNotReady
    status: "False"
    type: Ready
  daemonEndpoints:
    kubeletEndpoint:
      Port: 10250
  nodeInfo:
    architecture: amd64
    bootID: 89aac7ec-ca74-4d37-904e-74c6cff64fc2
    containerRuntimeVersion: docker://1.13.1
    kernelVersion: 3.10.0-862.11.6.el7.x86_64
    kubeProxyVersion: v1.10.7
    kubeletVersion: v1.10.7
    machineID: e51cc8a02c72433080ae4eb7d897f3eb
    operatingSystem: linux
    osImage: CentOS Linux 7 (Core)
    systemUUID: 8A04C2F2-AA29-469B-A9FB-52AC14FF97F7
```
</details>

<details>
  <summary>pod</summary>

```

$ k get pod -o yaml $(k get pods --field-selector=spec.nodeName==centos---1-10-7-hz-alvaro-5cb4687446-hhgd8 |grep flannel|awk '{ print $1}')
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: 2018-10-05T19:36:45Z
  generateName: kube-flannel-ds-amd64-
  labels:
    app: flannel
    controller-revision-hash: 6697bf5fc6
    pod-template-generation: "1"
    tier: node
  name: kube-flannel-ds-amd64-c4h24
  namespace: kube-system
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: DaemonSet
    name: kube-flannel-ds-amd64
    uid: a29fa9bf-c8d4-11e8-bffe-9600001198f6
  resourceVersion: "2645"
  selfLink: /api/v1/namespaces/kube-system/pods/kube-flannel-ds-amd64-c4h24
  uid: fe7c81c5-c8d5-11e8-bffe-9600001198f6
spec:
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchFields:
          - key: metadata.name
            operator: In
            values:
            - centos---1-10-7-hz-alvaro-5cb4687446-hhgd8
  containers:
  - args:
    - --ip-masq
    - --kube-subnet-mgr
    command:
    - /opt/bin/flanneld
    env:
    - name: POD_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
    - name: POD_NAMESPACE
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.namespace
    image: quay.io/coreos/flannel:v0.10.0-amd64
    imagePullPolicy: IfNotPresent
    name: kube-flannel
    resources:
      limits:
        cpu: 100m
        memory: 50Mi
      requests:
        cpu: 100m
        memory: 50Mi
    securityContext:
      privileged: true
      procMount: Default
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /run
      name: run
    - mountPath: /etc/kube-flannel/
      name: flannel-cfg
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: flannel-token-llwxk
      readOnly: true
  dnsPolicy: ClusterFirst
  hostNetwork: true
  initContainers:
  - args:
    - -f
    - /etc/kube-flannel/cni-conf.json
    - /etc/cni/net.d/10-flannel.conflist
    command:
    - cp
    image: quay.io/coreos/flannel:v0.10.0-amd64
    imagePullPolicy: IfNotPresent
    name: install-cni
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /etc/cni/net.d
      name: cni
    - mountPath: /etc/kube-flannel/
      name: flannel-cfg
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: flannel-token-llwxk
      readOnly: true
  nodeName: centos---1-10-7-hz-alvaro-5cb4687446-hhgd8
  nodeSelector:
    beta.kubernetes.io/arch: amd64
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: flannel
  serviceAccountName: flannel
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoSchedule
    operator: Exists
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
  - effect: NoSchedule
    key: node.kubernetes.io/disk-pressure
    operator: Exists
  - effect: NoSchedule
    key: node.kubernetes.io/memory-pressure
    operator: Exists
  - effect: NoSchedule
    key: node.kubernetes.io/unschedulable
    operator: Exists
  - effect: NoSchedule
    key: node.kubernetes.io/network-unavailable
    operator: Exists
  volumes:
  - hostPath:
      path: /run
      type: ""
    name: run
  - hostPath:
      path: /etc/cni/net.d
      type: ""
    name: cni
  - configMap:
      defaultMode: 420
      name: kube-flannel-cfg
    name: flannel-cfg
  - name: flannel-token-llwxk
    secret:
      defaultMode: 420
      secretName: flannel-token-llwxk
status:
  message: Pod Predicate MatchNodeSelector failed
  phase: Failed
  reason: MatchNodeSelector
  startTime: 2018-10-05T19:36:45Z
```

</details>

**Comment by k82cn** [comment]
https://github.com/kubernetes/kubernetes/pull/62453 is not cherry pick to 1.10; so the affinity filed can not be accepted by kubelet.

**Comment by k82cn** [comment]
AFAIK, we should upgrade from 1.10 to 1.11 than 1.12; if so, there'll be no issue :)

/cc @kubernetes/sig-cluster-lifecycle

**Comment by alvaroaleman** [comment]
@k82cn My understanding of the Kubernetes version skew policy has been that GA features should be supported for +2 minor releases. I would consider "Kubelet schedules DS pods" a GA feature.  @liggitt can you clarify?

What makes this worse from an operators perspective is that it is very hard to understand what causes the Kubelet do not schedule pods anymore. At the very least, this should be documented somewhere.

**Comment by neolit123** [comment]
> AFAIK, we should upgrade from 1.10 to 1.11 than 1.12; if so, there'll be no issue

+1

i can't find anywhere else where this is defined in the docs, but:
https://kubernetes.io/docs/setup/independent/install-kubeadm/#installing-kubeadm-kubelet-and-kubectl
> .. one minor version skew between the kubelet and the control plane is supported, but the kubelet version may never exceed the API server version

**Comment by liggitt** [comment]
> AFAIK, we should upgrade from 1.10 to 1.11 then 1.12; if so, there'll be no issue :)

The 1.12 control plane must support kubelets at levels 1.10 and 1.11. It cannot enable new features by default that break existing workloads for kubelets on those levels.

**Comment by liggitt** [comment]
This means that rolling 1.12 out to a working cluster with 1.10-level kubelets will break working daemonsets, correct?

**Comment by alvaroaleman** [comment]
> This means that rolling 1.12 out to a working cluster with 1.10-level kubelets will break working daemonsets, correct?

Yes, the Kubelet will remove them as it thinks that they are not supposed to run on it

**Comment by liggitt** [comment]
cc @kubernetes/sig-scheduling-bugs

**Comment by liggitt** [comment]
> @liggitt I think the APIserver is irrelevant here, relevant is the scheduler and controller-manager

yes, clarified

**Comment by neolit123** [comment]
i'd love to be corrected on that but i don't think there are any X-2 kubelet vs control-plane test jobs.

**Comment by k82cn** [comment]
> do we have CI scenarios that exercise workloads on a 1.10 kubelet against a 1.12 apiserver?

I think we do not have such a test :)


> The 1.12 control plane must support kubelets at levels 1.10 and 1.11.

Got that; prefer to give an known issues for that :)

**Comment by liggitt** [comment]
> prefer to give an known issues for that :)

I don't think that's sufficient. That's essentially a known issue of "kubelets are not actually supported at n-2 skew levels". We should add a known issue to inform people considering upgrading their control plane to 1.12, but also take action to restore the claimed support for skew levels.

**Comment by liggitt** [comment]
one possibility is to use the kubelet version reported by the node (`status.nodeInfo.kubeletVersion`) to determine whether to use a direct node assignment or matchfield affinity

**Comment by BenTheElder** [comment]
Testing checking in: I didn't think that we supported skewing that far, I don't think we do any tests against this intentionally because it's not supported. AFAIK components talk to each other skewed by one release, not two.

as @neolit123 called out, kubeadm specifically calls this out as well: https://github.com/kubernetes/kubernetes/issues/69346#issuecomment-427804831

**Comment by liggitt** [comment]
see https://github.com/kubernetes/community/blob/master/contributors/design-proposals/release/versioning.md#supported-releases-and-component-skew

> Nodes may lag masters components by up to two minor versions but should be at a version no newer than the master

that has been the kubelet/control-plane supported skew level since [~1.4](https://github.com/kubernetes/kubernetes/blob/v1.4.0/docs/design/versioning.md#supported-releases-and-component-skew)

At one point, I'm pretty sure we had tests that ran nodes skewed 2 versions older than apiservers. Are the gke-gci-1.9-gci-1.11-... jobs remnants of those?

**Comment by BenTheElder** [comment]
possibly. This ties into what I've been discussing with @spiffxp regarding test-ownership.. I don't know who maintains those .. :|

**Comment by neolit123** [comment]
sig-gcp for this whole tab:
https://k8s-testgrid.appspot.com/sig-cluster-lifecycle-upgrade-skew#Summary

> Nodes may lag masters components by up to two minor versions but should be at a version no newer than the master

AFAIK there was something specific about the `1.9-gci-1.11` which was testing gke/gci infrastructure and it was not a case where the test conforms the above claim.

@timothysc to confirm.

**Comment by BenTheElder** [comment]
Er right but even "sig-gcp" or  "sig-cluster-lifecycle" owns this is not terribly helpful compared to "these people actually know about and actively maintain this test" ... sig-gcp / sig-cluster-lifecycle encompass many things and I'm sure most can't keep up with all of that either.

I'm not sure who's actually the effective owner of these.

**Comment by neolit123** [comment]
https://github.com/kubernetes/test-infra/commit/a90445f25795a4ce51ae6d5b38e8ca104e2716da

**Comment by k82cn** [comment]
> one possibility is to use the kubelet version reported by the node (status.nodeInfo.kubeletVersion) to determine whether to use a direct node assignment or matchfield affinity

I'm going to create a PR for 1.12 to handle that if no other options; 1.13 is not necessary.

**Comment by liggitt** [comment]
resolved in 1.12.2 in https://github.com/kubernetes/kubernetes/pull/69566

note that before upgrading your control plane to 1.13, all kubelets in your cluster must be at 1.11 or newer
